### PR TITLE
Feature/arangobench custom query bind vars

### PR DESCRIPTION
**Pull request by cpjulia**
### Scope & Purpose
In arangobench, queries can be given as an argument with the flags `--custom-query` or `custom-query-file`, but they still do not support bind variables. This PR allows queries given in these flags to have bind variables by using the new flag `custom-query-bindvars`, which would have, as an argument, a json string with the bind variables and their values, e.g.
`--custom-query-bindvars '{"@collectionName": "testCollection", 'name': test}' --custom-query 'FOR doc in @@collectionName FILTER  doc.name = @name RETURN doc'`

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

No backports required

#### Related Information

- [x] Docs PR: https://github.com/arangodb/docs/pull/830
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -1,6 +1,9 @@
 devel
 -----
 
+* Added new flag `custom-query-bindvars` to arangobench, so queries given as 
+  argument with the flag `custom-query` can have bind variables in them.
```
Comment (jsteemann) [inline CHANGELOG:5]:
```suggestion
* Added new option `--custom-query-bindvars` to arangobench, so queries given via 
  option `--custom-query` can have bind variables in them.
```

File: client-tools/Benchmark/BenchFeature.cpp [hunk 2]
```diff
@@ -250,6 +257,18 @@ void BenchFeature::validateOptions(std::shared_ptr<ProgramOptions> options) {
           << _percentiles << "': histogram is disabled by default. Enable it with flag '--histogram.generate = true'.";
     }
   }
+  LOG_DEVEL << _customQuery;
```
Comment (jsteemann) [inline client-tools/Benchmark/BenchFeature.cpp:260]:
```suggestion
```

File: client-tools/Benchmark/BenchFeature.cpp [hunk 3]
```diff
@@ -250,6 +257,18 @@ void BenchFeature::validateOptions(std::shared_ptr<ProgramOptions> options) {
...
+  if (!_customQueryBindVars.empty()) {
+    _customQueryBindVarsBuilder = VPackParser::fromJson(_customQueryBindVars);
+  }
+
+  if(!options->processingResult().touched("--custom-query") && !options->processingResult().touched("--custom-query-file")) {
+    if(options->processingResult().touched("--custom-query-bindvars")) {
+      LOG_TOPIC("d06cf", WARN, arangodb::Logger::BENCH)
+          << "For flag '--custom-query-bindvars "
+          << _customQueryBindVars << "': custom query must be provided with flag '--custom-query' or 'custom-query-file'.";
```
Comment (jsteemann) [inline client-tools/Benchmark/BenchFeature.cpp:269]:
```suggestion
          << _customQueryBindVars << "': custom query must be provided with flag '--custom-query' or '--custom-query-file'.";
```

File: client-tools/Benchmark/BenchFeature.cpp [hunk 4]
```diff
@@ -250,6 +257,18 @@ void BenchFeature::validateOptions(std::shared_ptr<ProgramOptions> options) {
           << _percentiles << "': histogram is disabled by default. Enable it with flag '--histogram.generate = true'.";
     }
   }
+  LOG_DEVEL << _customQuery;
+  if (!_customQueryBindVars.empty()) {
+    _customQueryBindVarsBuilder = VPackParser::fromJson(_customQueryBindVars);
+  }
+
+  if(!options->processingResult().touched("--custom-query") && !options->processingResult().touched("--custom-query-file")) {
```
Comment (jsteemann) [inline client-tools/Benchmark/BenchFeature.cpp:265]: I think we can get away without this warning.

File: client-tools/Benchmark/BenchFeature.cpp [hunk 5]
```diff
@@ -250,6 +257,18 @@ void BenchFeature::validateOptions(std::shared_ptr<ProgramOptions> options) {
           << _percentiles << "': histogram is disabled by default. Enable it with flag '--histogram.generate = true'.";
     }
   }
+  LOG_DEVEL << _customQuery;
+  if (!_customQueryBindVars.empty()) {
+    _customQueryBindVarsBuilder = VPackParser::fromJson(_customQueryBindVars);
```
Comment (jsteemann) [inline client-tools/Benchmark/BenchFeature.cpp:262]: What happens if the specified JSON is invalid? Will the end user get a proper error message in that case?

File: client-tools/Benchmark/testcases/CustomQueryTest.h [hunk 6]
```diff
@@ -70,9 +72,13 @@ namespace arangodb::arangobench {
       using namespace arangodb::velocypack;
       requestData.payload.openObject();
       requestData.payload.add("query", Value(_query));
+      if (_queryBindVars != nullptr) {
+     //   LOG_DEVEL << _queryBindVars->slice().toString();
+        requestData.payload.add("bindVars", _queryBindVars->slice());
+      }
       requestData.payload.close();
+    //  LOG_DEVEL << requestData.payload.toString();
```
Comment (jsteemann) [inline client-tools/Benchmark/testcases/CustomQueryTest.h:80]:
```suggestion
      if (_queryBindVars != nullptr) {
        requestData.payload.add("bindVars", _queryBindVars->slice());
      }
      requestData.payload.close();
```

File: js/client/modules/@arangodb/testsuites/arangobench.js [hunk 7]
```diff
@@ -143,7 +143,37 @@ const benchTodos = [{
...
+  'threads': '1',
+  'test-case': 'version',
+  'keep-alive': 'true',
+  // test with Unicode database name
+  'server.database': '이것은 테스트입니까 ! @abc " mötör',
+  'create-database': true,
+  'collection': 'testCollection',
+  //these flags have double @s because of the feature that trims the @ for escaping it in configuration files in /etc
+  //the double @s will be removed when this feature is deprecated
+  'custom-query': 'FOR doc IN @@@@collectionName FILTER doc.name == @@name',
```
Comment (jsteemann) [inline js/client/modules/@arangodb/testsuites/arangobench.js:158]: This is not a valid AQL query.

File: js/client/modules/@arangodb/testsuites/arangobench.js [hunk 8]
```diff
@@ -143,7 +143,37 @@ const benchTodos = [{
...
```
Comment (jsteemann) [inline js/client/modules/@arangodb/testsuites/arangobench.js:172]: This is not a valid AQL query.

File: js/client/modules/@arangodb/testsuites/arangobench.js [hunk 9]
```diff
@@ -201,67 +231,85 @@ function arangobench (options) {
...
       }
-
       let oneResult = pu.run.arangoBenchmark(options, instanceInfo, args, instanceInfo.rootDir, options.coreCheck);
-      print();
-
-      if (benchTodo.hasOwnProperty('duration')) {
-        oneResult.status = oneResult.status && oneResult.duration >= benchTodo['duration'];
-        if (!oneResult.status) {
-          oneResult.message += ` didn't run for the expected time ${benchTodo.duration} but only ${oneResult.duration}`;
+      if (benchTodo.hasOwnProperty('expected-failure')) {
```
Comment (jsteemann) [inline js/client/modules/@arangodb/testsuites/arangobench.js:239]: This only checks if `expected-failure` is present, but not if it is set to `true` or `false`. This is a bit counter-intutitive, e.g. if someone else adds a test and sets the flag to `false`.

File: client-tools/Benchmark/BenchFeature.cpp (hunk 5 above)
Comment (cpjulia) [reply client-tools/Benchmark/BenchFeature.cpp]: Addressed

File: client-tools/Benchmark/BenchFeature.cpp (hunk 4 above)
Comment (cpjulia) [reply client-tools/Benchmark/BenchFeature.cpp]: Addressed

File: js/client/modules/@arangodb/testsuites/arangobench.js (hunk 8 above)
Comment (cpjulia) [reply js/client/modules/@arangodb/testsuites/arangobench.js]: Addressed

File: js/client/modules/@arangodb/testsuites/arangobench.js (hunk 9 above)
Comment (cpjulia) [reply js/client/modules/@arangodb/testsuites/arangobench.js]: Addressed

File: client-tools/Benchmark/BenchFeature.cpp [hunk 10]
```diff
@@ -250,6 +257,16 @@ void BenchFeature::validateOptions(std::shared_ptr<ProgramOptions> options) {
           << _percentiles << "': histogram is disabled by default. Enable it with flag '--histogram.generate = true'.";
     }
   }
+  if (!_customQueryBindVars.empty()) {
+    try {
+      _customQueryBindVarsBuilder = VPackParser::fromJson(_customQueryBindVars);
+    } catch (...) {
+      LOG_TOPIC("a3468", FATAL, arangodb::Logger::BENCH)
+          << "For flag '--custom-query-bindvars "
+          << _customQueryBindVars << "': invalid JSON format. ";
```
Comment (jsteemann) [inline client-tools/Benchmark/BenchFeature.cpp:266]:
```suggestion
          << _customQueryBindVars << "': invalid JSON format.";
```

**Review by jsteemann — APPROVED** [review]
LGTM!

**Pull request merged**